### PR TITLE
Return Non-zero Exit Code when Action Exceeds Time Limit

### PR DESCRIPTION
**Pull request by dubeejw**
- Return 202 exit code when an blocking action exceeds its time limit.

Closes: https://github.com/openwhisk/openwhisk/issues/1467

### CODE REVIEW

**Comment by dubeejw** [comment]
@mdeuser, this is definitely a WIP. Thoughts on this method of return the exit code? I think it could easily be translated over into the API portion of the CLI.

**Comment by dubeejw** [comment]
@mdeuser

**Comment by rabbah** [comment]
It's not the case that for all 202 the time limit was exceeded. It's only true for blocking invoke. Fwiw.

File: tools/cli/go-whisk-cli/commands/action.go [hunk 1]
```diff
@@ -166,7 +166,7 @@ var actionInvokeCmd = &cobra.Command{
     if err != nil {
       whiskErr, isWhiskErr := err.(*whisk.WskError)
 
-      if (isWhiskErr && whiskErr.ApplicationError != true) || !isWhiskErr {
+      if (isWhiskErr && whiskErr.ApplicationError != true && whiskErr.ExitCode != 202) || !isWhiskErr {
```
Comment (mdeuser) [inline tools/cli/go-whisk-cli/commands/action.go]: Special handling of 202 response codes should only happen on blocking action invocations

File: tools/cli/go-whisk/whisk/client.go [hunk 2]
```diff
@@ -293,7 +295,12 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
     // Handle 1. HTTP Success + Valid body matching request expectations
     // Handle 3. HTTP Success + Body does NOT match request expectations
     if IsHttpRespSuccess(resp) && v != nil {
-        return parseSuccessResponse(resp, data, v), nil
+        if resp.StatusCode == 202 {
+            err = MakeWskError(errors.New("Time limit exceeded."), 202, NO_DISPLAY_MSG, NO_DISPLAY_USAGE,
+                NO_MSG_DISPLAYED)
+        }
+
+        return parseSuccessResponse(resp, data, v), err
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/client.go]: Does a 202 response have a body to parse?

**Comment by mdeuser** [comment]
@dubeejw - the CLI already returns a few special exit codes that do not correspond to a HTTP status code, so another one can be defined for action invocation timeouts (i.e. EXITCODE_ERR_NOT_COMPLETED).

File: tools/cli/go-whisk/whisk/client.go (hunk 2 above)
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: @mdeuser, I can only speak for action invocations. They return only an activation ID.

**Comment by dubeejw** [comment]
@rabbah, from conversations with @mdeuser, it sounds like all of the API commands in the CLI will need to account for the 202 as well.

**Comment by dubeejw** [comment]
PG3 290.

**Comment by dubeejw** [comment]
PG2 1090.

**Comment by dubeejw** [comment]
@mdeuser, can you review this while I try to get a good PG?

File: tools/cli/go-whisk/whisk/wskerror.go [hunk 3]
```diff
@@ -36,6 +38,7 @@ type WskError struct {
     MsgDisplayed        bool    // When true, the error message has already been displayed, don't display it again
     DisplayUsage        bool    // When true, the CLI usage should be displayed before exiting
     ApplicationError    bool    // When true, the error is a result of an application failure
+    TimedOut            bool    // When True, the backend server responded with a 202 HTTP status code
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/wskerror.go]: Update function comment above to include the new parameter

File: tools/cli/go-whisk/whisk/client.go [hunk 4]
```diff
@@ -293,7 +297,12 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
     // Handle 1. HTTP Success + Valid body matching request expectations
     // Handle 3. HTTP Success + Body does NOT match request expectations
     if IsHttpRespSuccess(resp) && v != nil {
-        return parseSuccessResponse(resp, data, v), nil
+        if processTimeout && resp.StatusCode == EXITCODE_TIMED_OUT {
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/client.go]: Add comments here and above explaining timeout handling

File: tools/cli/go-whisk/whisk/client.go [hunk 5]
```diff
@@ -293,7 +297,12 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
...
+            err = MakeWskError(errors.New("Time limit exceeded."), EXITCODE_TIMED_OUT, NO_DISPLAY_MSG, NO_DISPLAY_USAGE,
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/client.go]: - Globalize message.
- I think this message will need to be a bit more descriptive.. That is, the corresponding request is still being processed and that the activation record must be queried to see the final results of the request... but in a more pithy message :-)
- Oh.. I see the `NO_DISPLAY_MSG` flag here.. Per above bullet, I think we need to tell the user something more than just dumping the activation json object..

File: tools/cli/go-whisk/whisk/action.go [hunk 6]
```diff
@@ -203,7 +203,7 @@ func (s *ActionService) Invoke(actionName string, payload interface{}, blocking
         return nil, nil, whiskErr
     }
 
-    resp, err := s.client.Do(req, &res)
+    resp, err := s.client.Do(req, &res, blocking)
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/action.go]: Perhaps more clear if the designated "timeout handling" constant was used.

File: tools/cli/go-whisk/whisk/api.go [hunk 7]
```diff
@@ -152,7 +152,7 @@ func (s *ApiService) List(apiListOptions *ApiListOptions) (*RetApiArray, *http.R
     }
 
     apiArray := new(RetApiArray)
-    resp, err := s.client.Do(req, &apiArray)
+    resp, err := s.client.Do(req, &apiArray, DoNotProcessTimeOut)
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/api.go]: All of the api requests to the controller are actually action invocations, so all of the Do() calls in this module need the ExitWithErrorOnTimeout/ProcessTimeout constant.

**Comment by dubeejw** [comment]
@mdeuser, @csantanapr, @rabbah, I'm sure there will be opinions on the output of a blocking invocation that has gone async. Here's what I currently have:
 
```
 wsk action invoke time -b
ok: invoked /_/time with id 6b239df03aaa469fb64f3d2cc57af267
{
    "activationId": "6b239df03aaa469fb64f3d2cc57af267"
}
error: Request accepted, but processing not completed yet.
```

**Comment by rabbah** [comment]
The wording is ok but I'm not in favor of reporting this as an error because it's not consistent with the API response. 20x is success and errors from the controller are accompanied by both the error message and the transaction code.

**Comment by dubeejw** [comment]
PG3 312, approved.

**Comment by dubeejw** [comment]
Output from a blocking invocation that exceeds its time limit:
```
wsk action invoke time -b
ok: invoked /_/time with id 677ccf2ce2264c3e8701e0477c6c804e
{
    "activationId": "677ccf2ce2264c3e8701e0477c6c804e"
}
Request accepted, but processing not completed yet.
```

**Comment by rabbah** [comment]
should we just have one line?

`ok: invoked /_/time but the request has not yet finished, activation id xyz`

**Comment by dubeejw** [comment]
```
wsk action invoke time -b
ok: invoked /time/_ with id 83d442d30dae4749b5517e7cf6d10e9d, but the request has yet to finish
```

**Comment by rabbah** [comment]
i like it but you might want to check the parsing that happens in the Wsk helpers, i think they assume the id is at the end... travis will tell us i suppose.

otherwise looks good!!

**Comment by csantanapr** [comment]
@dubeejw Travis just told us
```
system.basic.WskSequenceTests > Wsk Sequence should execute a sequence in blocking fashion and finish execution even if longer than blocking response timeout FAILED
    org.scalatest.exceptions.TestFailedException: did not find an activation id in 'exit code = 202
    stdout: 
    stderr: ok: invoked /sSequence/_ with id 29320f3650b3415a91c7a96d14d7628e, but the request has yet to finish
    ' None was not an instance of scala.Some, but an instance of scala.None$
```

**Comment by csantanapr** [comment]
Hi @dubeejw sorry for the trouble but with this recent update 2861cf02cb8c6cd5cf60a94db1eb00812c283515

I see a conflict and it's not trivial to resolve.
Do you mind submitting a commit to fix the conflict.

**Comment by dubeejw** [comment]
@csantanapr, rebased. Still need a clean PG.

**Comment by dubeejw** [comment]
PG2 1126, approved.

**Comment by dubeejw** [comment]
Ready for merge, @rabbah?

**Comment by csantanapr** [comment]
going to merge this one

**Comment by csantanapr** [comment]
merged b581cc43a87e25cd4774313718f50fabca1efecc

**Pull request closed without merging**